homeassistant/components/miele/sensor.py

```diff
@@ -576,51 +592,92 @@ async def async_setup_entry(
+            # don't create entity if API signals that datapoint is disabled, unless the sensor
+            # already appeared in the past (= it provided a valid value)
+            return _is_entity_registered(unique_id)
+        if (
+            definition.description.key == "state_plate_step"
+            and definition.description.zone is not None
+            and definition.description.zone > _get_plate_count(device.tech_type)
+        ):
+            # don't create plate entity if not expected by the appliance tech type
+            return False
+        return True
 
-    def _async_add_new_devices() -> None:
-        nonlocal added_devices
+    def _async_add_devices() -> None:
+        nonlocal added_devices, added_entities
         entities: list = []
         entity_class: type[MieleSensor]
         new_devices_set, current_devices = coordinator.async_add_devices(added_devices)
         added_devices = current_devices
 
         for device_id, device in coordinator.data.devices.items():
             for definition in SENSOR_TYPES:
-                if (
-                    device_id in new_devices_set
-                    and device.device_type in definition.types
-                ):
-                    match definition.description.key:
-                        case "state_status":
-                            entity_class = MieleStatusSensor
-                        case "state_program_id":
-                            entity_class = MieleProgramIdSensor
-                        case "state_program_phase":
-                            entity_class = MielePhaseSensor
-                        case "state_plate_step":
-                            entity_class = MielePlateSensor
-                        case _:
-                            entity_class = MieleSensor
-                    if (
-                        definition.description.device_class
-                        == SensorDeviceClass.TEMPERATURE
-                        and definition.description.value_fn(device)
-                        == DISABLED_TEMPERATURE / 100
-                    ) or (
-                        definition.description.key == "state_plate_step"
-                        and definition.description.zone
-                        > _get_plate_count(device.tech_type)
-                    ):
-                        # Don't create entity if API signals that datapoint is disabled
-                        continue
-                    entities.append(
-                        entity_class(coordinator, device_id, definition.description)
+                # device is not supported, skip
+                if device.device_type not in definition.types:
+                    continue
+
+                entity_class = _get_entity_class(definition)
+                unique_id = (
+                    definition.description.unique_id_fn(
+                        device_id, definition.description
```

Comment:
Okay so one thing is, we should only check this on creation, not every time we get an update, do we keep track of all data we already kept track off?

Reply:
> we should only check this on creation, not every time we get an update

But in this way, I would recognize new temperature sensors only at creation. We need that the temperature sensor is created when it provides a valid value for the first time because of API limitations.

> do we keep track of all data we already kept track off?

Sorry, I don't get the point here :)

Reply:
But you mentioned you don't know what values are available at boot right? But do we make sure we don't check every datapoint at every update?

Reply:
Yes, the API is always reporting something, valid or not valid value. The first time a temperature becomes valid, we need to create the sensor and keep it afterwards, reporting "unknown" as the API gets back to the invalid state (appliance is off or idle for example).
For this reason, at every update we need to check whether the sensor is enabled by looking at the value currently reported by API or in the registry.
This doesn't seem different than the previous implementation or I'm missing something.